highlight/src/lib.rs

```diff
@@ -0,0 +1,833 @@
+        // But because the tree is really a pointer to the heap, the cursor can remain
+        // valid when the tree is moved. There's no way to express this with lifetimes
+        // right now, so we have to `transmute` the cursor's lifetime.
+        let cursor = unsafe { transmute(tree.walk_with_properties(sheet, source)) };
+        Self {
+            _tree: tree,
+            cursor,
+            ranges,
+            at_node_end: false,
+        }
+    }
+
+    fn offset(&self) -> usize {
+        if self.at_node_end {
+            self.cursor.node().end_byte()
+        } else {
+            self.cursor.node().start_byte()
+        }
+    }
+
+    fn advance(&mut self) -> bool {
+        if self.at_node_end {
+            if self.cursor.goto_next_sibling() {
+                self.at_node_end = false;
+            } else if !self.cursor.goto_parent() {
+                return false;
+            }
+        } else if !self.cursor.goto_first_child() {
+            self.at_node_end = true;
+        }
+        true
+    }
+}
+
+impl<'de> Deserialize<'de> for Scope {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let s = String::deserialize(deserializer)?;
+        match s.as_str() {
+            "attribute" => Ok(Scope::Attribute),
+            "comment" => Ok(Scope::Comment),
+            "constant" => Ok(Scope::Constant),
+            "constant.builtin" => Ok(Scope::ConstantBuiltin),
+            "constructor" => Ok(Scope::Constructor),
+            "constructor.builtin" => Ok(Scope::ConstructorBuiltin),
+            "embedded" => Ok(Scope::Embedded),
+            "escape" => Ok(Scope::Escape),
+            "function" => Ok(Scope::Function),
+            "function.builtin" => Ok(Scope::FunctionBuiltin),
+            "keyword" => Ok(Scope::Keyword),
+            "number" => Ok(Scope::Number),
+            "operator" => Ok(Scope::Operator),
+            "property" => Ok(Scope::Property),
+            "property.builtin" => Ok(Scope::PropertyBuiltin),
+            "punctuation" => Ok(Scope::Punctuation),
+            "punctuation.bracket" => Ok(Scope::PunctuationBracket),
+            "punctuation.delimiter" => Ok(Scope::PunctuationDelimiter),
+            "punctuation.special" => Ok(Scope::PunctuationSpecial),
+            "string" => Ok(Scope::String),
+            "string.special" => Ok(Scope::StringSpecial),
+            "type" => Ok(Scope::Type),
+            "type.builtin" => Ok(Scope::TypeBuiltin),
+            "variable" => Ok(Scope::Variable),
+            "variable.builtin" => Ok(Scope::VariableBuiltin),
+            "tag" => Ok(Scope::Tag),
+            _ => Ok(Scope::Unknown),
+        }
+    }
+}
+
+pub trait HTMLAttributeCallback<'a>: Fn(Scope) -> &'a str {}
+
+pub fn highlight<'a, F>(
+    source: &'a [u8],
+    language: Language,
+    property_sheet: &'a PropertySheet<Properties>,
+    injection_callback: &'a F,
```

Comment:
The highlighting functionality is mostly exposed through this `highlight` function and the `highlight_html` wrapper.

They both take an `injection_callback` parameter that specifies how to retrieve other 'embedded' languages. This callback is necessary because that language retrieval works differently in different contexts. In the Tree-sitter CLI, the languages are loaded lazily and dynamically using `dlopen`, whereas in other apps, the languages may be statically linked into the app.